pkg/replica/hash.go

```diff
@@ -0,0 +1,413 @@
+				// The deletion failure is acceptable, because the mismatching timestamps
+				// will trigger the rehash in the next hash request.
+				DeleteSnapshotHashInfoFromXattr(t.SnapshotName)
+			}
+		}
+
+		if err != nil {
+			logrus.Errorf("failed to hash snapshot %v since %v", t.SnapshotName, err)
+			t.State = ProgressStateError
+			t.Error = err.Error()
+		}
+	}()
+
+	// Each node can have only one snapshot hashing task at the same time per node.
+	// When the snapshot hashing task is started, the task tries to fetch the file lock
+	// (${fileLockDirectory}/hash). If the lock file is held by another task, it will stuck
+	// here and wait for the lock. The file is unlocked after the task is completed.
+	fileLock, err := t.LockFile()
+	if err != nil {
+		return
+	}
+	defer t.UnlockFile(fileLock)
+
+	modTime, err = GetSnapshotModTime(t.SnapshotName)
+	if err != nil {
+		return
+	}
+
+	// If the silent corruption is detected, don't need to recalculate the checksum.
+	// Just set SilentlyCorrupted to true and return it.
+	isSilentlyCorrupted, err := t.isSilentCorruptionAlreadyDetected(modTime)
+	if err != nil {
+		return
+	} else if isSilentlyCorrupted {
+		t.StatusLock.Lock()
+		t.SilentlyCorrupted = true
+		t.StatusLock.Unlock()
+		return
+	}
+
+	requireRehash := true
+	if !t.Rehash {
+		requireRehash, checksum, err = t.isRehashRequired(modTime)
+		if err != nil {
+			return
+		}
+		if !requireRehash {
+			return
+		}
+	}
+
+	logrus.Infof("Starting hashing snapshot %v", t.SnapshotName)
+
+	lastHashedAt = time.Now().UTC().Format(time.RFC3339)
+	checksum, err = hashSnapshot(t.Ctx, t.SnapshotName)
+	if err != nil {
+		return
+	}
+
+	// If the silent corruption is detected, the xattr will not be overrode.
+	// The scene will be preserved and only set silentlyCorrupted to true.
+	if t.isSnapshotSilentlyCorrupted(checksum) {
+		silentlyCorrupted = true
+
+		info, err := GetSnapshotHashInfoFromXattr(t.SnapshotName)
+		if err != nil && err != syscall.ENODATA {
+			return
+		}
+
+		if info == nil {
+			checksum = info.Checksum
+			lastHashedAt = info.LastHashedAt
+			modTime = info.ModTime
+		}
+	}
+}
+
+func (t *SnapshotHashJob) isSnapshotSilentlyCorrupted(checksum string) bool {
+	// To detect the silent corruption, read the modTime and checksum already recorded in the snapshot disk file first.
+	// Then, rehash the file and compare the modTimes and checksums.
+	// If the modTimes are identical but the checksums differ, the file is silently corrupted.
+
+	info, err := GetSnapshotHashInfoFromXattr(t.SnapshotName)
+	if err != nil || info == nil {
+		return false
+	}
+
+	existingChecksum := info.Checksum
+	existingModTime := info.ModTime
+
+	if existingChecksum == "" || existingModTime == "" {
+		return false
+	}
+
+	if err := t.isModTimeRemain(existingModTime); err != nil {
+		return false
+	}
+
+	if checksum != existingChecksum {
+		return true
+	}
+
+	return false
+}
+
+func GetSnapshotModTime(snapshotName string) (string, error) {
```

Comment:
Again, if a function/method used internally, no need to export, so it should `getSnapshotModTime`. This way can make the code self-explanatory.

Reply:
`GetSnapshotModTime` is used by the functions in `rpc` and `replica` use

Reply:
Then ok!